// @chakra-ui/styled-system/dist/cjs/config/position.js
"use strict";

exports.__esModule = true;
exports.positionParser = exports.position = void 0;

var _core = require("../core");

var _utils = require("../utils");

var config = {
  position: true,
  pos: _utils.t.prop("position"),
  zIndex: {
    property: "zIndex",
    scale: "zIndices"
  },
  inset: _utils.t.spaceT("inset"),
  insetX: _utils.t.spaceT(["left", "right"]),
  insetInline: _utils.t.spaceT("insetInline"),
  insetY: _utils.t.spaceT(["top", "bottom"]),
  insetBlock: _utils.t.spaceT("insetBlock"),
  top: _utils.t.spaceT("top"),
  insetBlockStart: _utils.t.spaceT("insetBlockStart"),
  bottom: _utils.t.spaceT("bottom"),
  insetBlockEnd: _utils.t.spaceT("insetBlockEnd"),
  left: _utils.t.spaceT("left"),
  insetInlineStart: _utils.t.spaceT("insetInlineStart"),
  right: _utils.t.spaceT("right"),
  insetInlineEnd: _utils.t.spaceT("insetInlineEnd")
};
Object.assign(config, {
  insetStart: config.insetInlineStart,
  insetEnd: config.insetInlineEnd
});
/**
 * Types for position CSS properties
 */

var position = (0, _core.system)(config);
exports.position = position;
var positionParser = (0, _core.createParser)(config);
exports.positionParser = positionParser;
